### ISSUE TITLE
Negative justified epoch on first epoch transition

### ISSUE THREAD

**Issue by arnetheduck** [ethereum/eth2.0-specs#849]
This logic looks a bit fishy:

```
    if previous_boundary_attesting_balance * 3 >= get_previous_total_balance(state) * 2:
        new_justified_epoch = get_current_epoch(state) - 1
        state.justification_bitfield |= 2
...

def get_previous_total_balance(state: BeaconState) -> Gwei:
    return get_total_balance(state, get_active_validator_indices(state.validator_registry, get_previous_epoch(state)))

```

basically, during the first epoch transition, `get_previous_epoch`, if we weren't using the ugly integers, would be `-1` (because `state.slot` is updated after epoch processing). in particular, it means that there are no active validators because their activation epoch is set to `GENESIS_EPOCH`, and therefore the total attesting balance is 0. 

As a consequence, `state.current_justified_epoch` is set to `-1` - before the genesis epoch.

**Comment by arnetheduck** [comment]
Some more flavor:

@mkalinin noted that the next section resets the justified slot to 0:
```
    if current_boundary_attesting_balance * 3 >= get_current_total_balance(state) * 2:
        new_justified_epoch = get_current_epoch(state)
        state.justification_bitfield |= 1
```

However, this is true only iff there are enough attestations - we're using `get_current_total_balance` in this check which is not affected by the underflow - also, during the first epoch, there are naturally fewer attestations around because of `MIN_ATTESTATION_INCLUSION_DELAY` so the practical threshold is actually higher than 2/3's to enter here.

a temporary fix we're using is to saturate `get_previous_epoch` - but it has only received minimal testing:
```
return max(GENESIS_EPOCH, epoch - 1)
```

**Comment by djrtwo** [comment]
This is addressed in #896 by blocking `update_justification_and_finalization` on the first 2 epochs. prevents `-1` and resetting to `0`

Will close then that PR is merged

**Comment by djrtwo** [comment]
closed via #896
